tsdb/exemplar.go

```diff
@@ -398,7 +443,25 @@ func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, e exemplar.Exemp
 	ce.exemplars[ce.nextIndex].next = noExemplar
 	ce.exemplars[ce.nextIndex].exemplar = e
 	ce.exemplars[ce.nextIndex].ref = ce.index[string(seriesLabels)]
-	ce.index[string(seriesLabels)].newest = ce.nextIndex
+	// Only set newest if not an out of order exemplar
+	if insertionPoint == -1 {
+		ce.index[string(seriesLabels)].newest = ce.nextIndex
+	} else {
+		// The exemplar is out of order
+		prevEntry := ce.exemplars[insertionPoint]
+		idx := prevEntry.ref
+		oldest := ce.exemplars[idx.oldest].exemplar
+		if e.Ts < oldest.Ts || (e.Ts == oldest.Ts &&
+			(e.Value < oldest.Value || (e.Value == oldest.Value && e.Labels.Hash() < oldest.Labels.Hash()))) {
+			// The exemplar should be inserted before the oldest one
+			ce.exemplars[ce.nextIndex].next = ce.nextIndex
```

Comment:
isn't this a circular reference? shouldn't the newly inserted ooo exemplar point to the previously oldest exemplar?